apps/s_client.c

```diff
@@ -3025,15 +3025,22 @@ int s_client_main(int argc, char **argv)
     if (in_init)
         print_stuff(bio_c_out, con, full_log);
     do_ssl_shutdown(con);
-#if defined(OPENSSL_SYS_WINDOWS)
     /*
-     * Give the socket time to send its last data before we close it.
-     * No amount of setting SO_LINGER etc on the socket seems to persuade
-     * Windows to send the data before closing the socket...but sleeping
-     * for a short time seems to do it (units in ms)
-     * TODO: Find a better way to do this
+     * If we ended with an alert being sent, but still with data in the
+     * network buffer to be read, then calling BIO_closesocket() will
+     * result in a TCP-RST being sent. On some platforms (notably
+     * Windows) then this will result in the peer immediately abandoning
+     * the connection including any buffered alert data before it has
+     * had a chance to be read. Shutting down the sending side first,
+     * and then closing the socket sends TCP-FIN first followed by
+     * TCP-RST. This seems to allow the peer to read the alert data.
      */
-    Sleep(50);
+#ifdef _WIN32
+# ifdef SD_SEND
+    shutdown(SSL_get_fd(con), SD_SEND);
+# endif
+#elif defined(SHUT_WR)
+    shutdown(SSL_get_fd(con), SHUT_WR);
 #endif
     BIO_closesocket(SSL_get_fd(con));
```

Comment:
Actually, you can remove all that shutdown business *AND* the `BIO_closesocket()` call.  See, the `SSL_free()` call further down calls `BIO_free_all()`, which should end up having `conn_free`, and consequently `conn_close_socket`, which does the `shutdown` and `BIO_closesocket` calls correctly and in the right order.  So basically, the `BIO_closesocket` call here is wrong, and I've no idea why it's here.

Reply:
yes, good idea.

Reply:
Hmm....
Unfortunately `conn_close_socket` uses `shutdown(bio->num, 2)`
which means `SHUT_RDWR`, but I need the sequence SHUT_WR
followed by close or SHUT_RD.
So I don't think that will work.

Reply:
Actually, I think the *big* trouble here is that `conn_close_socket` isn't called at all, because this isn't a `bss_conn` BIO, it's a `bss_sock` one, and if you look in `crypto/bio/bss_sock.c`, you'll find that `sock_free` doesn't call `shutdown` *at* *all*...  I'm currently testing with an added `shutdown(a->num, 2)` to see if that makes a difference...

Reply:
funny thing is that the s_client does not even close the socket without the `BIO_closesocket`

```
17894 write(3, "\27\3\3\0\26test\n\27\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 27 <unfinished ...>
17894 <... write resumed> )             = 27
17894 select(4, [0 3], [], NULL, NULL <unfinished ...>
17894 <... select resumed> )            = 1 (in [0])
17894 read(0,  <unfinished ...>
17894 <... read resumed> "", 8192)      = 0
17894 write(2, "DONE\n", 5 <unfinished ...>
17894 <... write resumed> )             = 5
17894 write(3, "\27\3\3\0\23\1\0\25\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 24 <unfinished ...>
17894 <... write resumed> )             = 24
17894 exit_group(0)                     = ?
17894 +++ exited with 0 +++
```

Reply:
As I said, 2 = SHUT_RDWR, and in my theory that won't send the pending data,
because it is obvious that the received data won't be read any more.
to get the right effect, you need 1 = SHUT_WR, but I doubt we want to fix that
in bss_sock.c, as @kroeckx pointed out, it is a quite radical fix.

Reply:
Regarding s_client not even closing the socket without `BIO_closesocket`, that's because we're creating the BIO with `BIO_NOCLOSE`, and I can't quite figure out why we do it like that...  it seems a bit silly to have to duplicate in s_client the stuff that should happen in the BIO implementations...

Reply:
Note that the s_server does exactly the same, in fact I copied the code from s_socket.c

Reply:
I wonder if a two phase-shutdown of the client socket would read be a better option,
but I personally would be happy with a quick-and-simple fix that makes the test suite happy.

Reply:
I'm not so sure about quick-n-easy in this case, as it may be that we're uncovering bugs elsewhere (such as bss_sock.c)...  but I'm still now trying a testloop with just your patch.